### PR TITLE
reimplement all Bucket config APIs

### PR DESCRIPTION
**Pull request by harshavardhana**

### CODE REVIEW

**Comment by harshavardhana** [comment]
I removed ReplicationReqOpts on purpose as they are not idiomatic for all our existing APIs for buckets, any new options should be sent as part of replicationConfig instead of opts for the request.

**Comment by harshavardhana** [comment]
> I removed ReplicationReqOpts on purpose as they are not idiomatic for all our existing APIs for buckets, any new options should be sent as part of replicationConfig instead of opts for the request.

Either we add `Options` in all APIs or we don't add it right now and bring in when its needed.

File: api-bucket-policy.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+		contentLength: int64(len(b)),
+	}
+
+	// Execute PUT to upload a new bucket policy.
+	resp, err := c.executeMethod(ctx, "PUT", reqMetadata)
+	defer closeResponse(resp)
+	if err != nil {
+		return err
+	}
+	if resp != nil {
```
Comment (vadmeste) [inline api-bucket-policy.go:77]: if resp != nil doesn't seem to be needed

File: api-bucket-policy.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+	// using them in http request.
+	urlValues := make(url.Values)
+	urlValues.Set("policy", "")
+
+	// Content-length is mandatory for put policy request
+	policyReader := strings.NewReader(policy)
+	b, err := ioutil.ReadAll(policyReader)
+	if err != nil {
+		return err
+	}
```
Comment (vadmeste) [inline api-bucket-policy.go:62]: I don't think this block is needed, besides we should read from policyReader and then use it again.

File: api-bucket-policy.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+	reqMetadata := requestMetadata{
+		bucketName:    bucketName,
+		queryValues:   urlValues,
+		contentBody:   policyReader,
```
Comment (vadmeste) [inline api-bucket-policy.go:67]: strings.NewReader(policy)

File: api-bucket-policy.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+		contentLength: int64(len(b)),
```
Comment (vadmeste) [inline api-bucket-policy.go:68]: len(policy)

File: api-bucket-replication.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+	replication, err := xml.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+	// Content-length is mandatory for put replication request
+	replicationReader := bytes.NewReader(replication)
+	b, err := ioutil.ReadAll(replicationReader)
+	if err != nil {
+		return err
+	}
```
Comment (vadmeste) [inline api-bucket-replication.go:72]: This block doesn't seem to be needed

File: api-bucket-replication.go [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+	reqMetadata := requestMetadata{
+		bucketName:       bucketName,
+		queryValues:      urlValues,
+		contentBody:      replicationReader,
```
Comment (vadmeste) [inline api-bucket-replication.go:76]: bytes.NewReader(replication)

File: api-bucket-replication.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+		contentLength:    int64(len(b)),
```
Comment (vadmeste) [inline api-bucket-replication.go:77]: len(replication)

File: pkg/notification/info.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
+package notification
```
Comment (vadmeste) [inline pkg/notification/info.go:1]: Licence header is missing

File: api-bucket-policy.go [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+	"net/url"
+	"strings"
+
+	"github.com/minio/minio-go/v7/pkg/s3utils"
+)
+
+// SetBucketPolicy sets the access permissions on an existing bucket.
+func (c Client) SetBucketPolicy(ctx context.Context, bucketName, policy string) error {
+	// Input validation.
+	if err := s3utils.CheckValidBucketName(bucketName); err != nil {
```
Comment (kannappanr) [inline api-bucket-policy.go:32]: This check is not required as it is checked in both `removeBucketPolicy` and `putBucketPolicy`

File: api-bucket-replication.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+// RemoveBucketReplication removes a replication config on an existing bucket.
+func (c Client) RemoveBucketReplication(ctx context.Context, bucketName string) error {
+	return c.removeBucketReplication(ctx, bucketName)
+}
+
+// SetBucketReplication sets a replication config on an existing bucket.
+func (c Client) SetBucketReplication(ctx context.Context, bucketName string, cfg replication.Config) error {
+	// Input validation.
+	if err := s3utils.CheckValidBucketName(bucketName); err != nil {
```
Comment (kannappanr) [inline api-bucket-replication.go:40]: This check is also not required

File: pkg/sse/sse.go [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+	}
+}
+
+// NewConfigurationSSEKMS initializes a new SSE-KMS configuration
+func NewConfigurationSSEKMS(kmsMasterKey string) *Configuration {
+	return &Configuration{
+		Rules: []Rule{
+			{
+				Apply: ApplySSEByDefault{
+					KmsMasterKeyID: "my-masterkey",
```
Comment (kannappanr) [inline pkg/sse/sse.go:60]:
```suggestion
					KmsMasterKeyID: kmsMasterKey,
```

[the remark this answers is not in the document]
Comment (harshavardhana) [reply pkg/sse/sse.go:60]: No it is kmsMasterKeyId @vadmeste

File: api-bucket-policy.go (hunk 1 above)
Comment (harshavardhana) [reply api-bucket-policy.go:77]: old code @vadmeste not changing it.

File: api-bucket-policy.go (hunk 2 above)
Comment (harshavardhana) [reply api-bucket-policy.go:62]: old code @vadmeste not changing it right now

File: pkg/notification/info.go (hunk 8 above)
Comment (kannappanr) [inline pkg/notification/info.go:1]: copyright info missing

File: api-lifecycle.go [hunk 12]
```diff
@@ -0,0 +1,156 @@
+/*
+ * MinIO Go Library for Amazon S3 Compatible Cloud Storage
+ * Copyright 2015-2017 MinIO, Inc.
```
Comment (kannappanr) [inline api-lifecycle.go:3]:
```suggestion
 * Copyright 2020 MinIO, Inc.
```

File: api-lifecycle.go [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+	"net/url"
+
+	"github.com/minio/minio-go/v7/pkg/lifecycle"
+	"github.com/minio/minio-go/v7/pkg/s3utils"
+)
+
+// SetBucketLifecycle set the lifecycle on an existing bucket.
+func (c Client) SetBucketLifecycle(ctx context.Context, bucketName string, config *lifecycle.Configuration) error {
+	// Input validation.
+	if err := s3utils.CheckValidBucketName(bucketName); err != nil {
```
Comment (kannappanr) [inline api-lifecycle.go:35]: This check is not required. Otherwise the checks in removeBucketLifecyle and putBucketLifecylce should be removed

**Comment by harshavardhana** [comment]
Addressed all comments @kannappanr @vadmeste @poornas

**Pull request merged**